### PR TITLE
Date Picker Dialog Example: Improve high contrast support and refactor javascript

### LINKED ISSUE #1360: Update date picker dialog example to use single object and remove buttons from date grid
* Improve code to use a single object
* Remove buttons from the grid of dates
* Make sure that `aria-select` only set on the current date
* Check to make sure "space" updates the selected date and does not close dialog

### PR DESCRIPTION
**Pull request by jongund**
Resolves #1360 with the following changes:
* Updates Javascript to use a single object.
* Fixes space key issues in grid.
* Improves high contrast support.

#### Preview link

[View changes in the compare branch for this PR](https://raw.githack.com/w3c/aria-practices/issue1360-update-datepicker-dialog-example/examples/dialog-modal/datepicker-dialog.html)

### Review checklist

* [x] [Editorial review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#editorial_review) by @mcking65 @carmacleod 
* [x] [Functional review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#functional_review) by @carmacleod
* [x] [Visual design review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#design_review) by @a11ydoer 
* [ ] [Accessibility review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#accessibility_review) by @carmacleod
* [x] [Code review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#code_review) by @zcorpan
* [x] [Test review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#test_review) by @zcorpan and @spectranaut

### CODE REVIEW

**Comment by jongund** [comment]
@spectranaut @smhigley @mcking65 

I updated the regression tests to fix the SPACE key issue in Issue #1358

**Comment by zcorpan** [comment]
Part of the regression test for the date picker is very slow to run. It's focusing each cell with half a second delay between each step. Do you know why it's slow?

```
$ npm run regression -- --match "*datepicker*"

> aria-practices@0.0.0 regression /Users/zcorpan/git/w3c/aria-practices
> ava --timeout=1m "--match" "*datepicker*"



  43 tests passed
  2 known failures

  dialog-modal_datepicker › dialog-modal/datepicker-dialog.html [data-test-id="month-year-button-space-return"]: SPACE to buttons change calendar and date in focus
  dialog-modal_datepicker › dialog-modal/datepicker-dialog.html [data-test-id="grid-space-return"]: SPACE or RETURN selects date in focus
```
Are the known failures intended to be failures still?

File: test/tests/dialog-modal_datepicker.js [hunk 1]
```diff
@@ -696,7 +706,7 @@ ariaTest('SPACE on cancel button does not select date', exampleFile, 'okay-cance
 
   await setDateToJanFirst2019(t);
   await t.context.session.findElement(By.css(ex.buttonSelector)).click();
-  await t.context.session.findElement(By.css(ex.currentlyFocusedButton)).sendKeys(Key.ARROW_RIGHT);
+  await t.context.session.findElement(By.css(ex.currentlyFocusedDay)).sendKeys(Key.ARROW_RIGHT);
   await t.context.session.findElement(By.css(ex.cancelButton)).sendKeys(Key.SPACE);
```
Comment (zcorpan) [inline test/tests/dialog-modal_datepicker.js:710]: Should all `Key.SPACE`s be replaced with `' '`? Maybe we can lint it also?

**Comment by zcorpan** [comment]
> Improved high contrast support

The focus indicator is a 2px white border instead of a 1px white border. Could it be a different color, or more visible in some other way?

Screenshot of the date picker with focus on a date's cell.

**Comment by jongund** [comment]
@zcorpan Your right there is not much difference, I am going to change the border to be dashed to  see if it stands out more.

File: test/tests/dialog-modal_datepicker.js (hunk 1 above)
Comment (jongund) [reply test/tests/dialog-modal_datepicker.js:710]: Sounds like a good idea.

Comment (jongund) [reply test/tests/dialog-modal_datepicker.js:710]: @zcorpan 

I just did some testing and we can only do this when an keyboard event handler is associated with a button.  So I think we should only use `" " ` technique when needed and use `Key.SPACE` when it works correctly,

**Comment by jongund** [comment]
@zcorpan 

I made some updates for high contrast support for dates and buttons in the dialog, can you review and comment on the changes.

**Comment by jongund** [comment]
@zcorpan 
@smhigley 
I merged the changes to the regression test and there are two failing tests now, similar to the failures in Pull 1017.   How do I fix these tests?

```
2 tests failed

  dialog-modal_datepicker » dialog-modal/datepicker-dialog.html [data-test-id="gridcell-button-tabindex"]: Roving tab index on dates in gridcell

  node_modules\selenium-webdriver\lib\error.js:550

  Rejected promise returned by test. Reason:

  InvalidArgumentError {
    remoteStacktrace: '',
    message: 'missing field `value` at line 1 column 24',
  }



  dialog-modal_datepicker » dialog-modal/datepicker-dialog.html [data-test-id="gridcell-button-aria-selected"]: aria-selected on selected date

  test\util\queryElements.js:20

   19:   if (result.length === 0) {
   20:     t.fail(`Element query returned no results: ${selector}`);
   21:   }

  Element query returned no results: undefined[aria-selected="true"]
```

**Comment by jongund** [comment]
@mcking65 
@carmacleod 
Updated the documentation on the high contrast for the dialog buttons and dates.

**Comment by zcorpan** [comment]
The properties in the `ex` definition was changed, but the tests still used the old `allDateButtons` property. This caused the selector `undefined[aria-selected="true"]`. Fixed in e3c7836

**Comment by zcorpan** [comment]
I've changed the remaining `findElement` to `queryElements`. I've also added missing `await`s and documented this subtle mistake [in the wiki](https://github.com/w3c/aria-practices/wiki/Regression-Tests-for-APG-Example-Pages#common-mistakes).

**Comment by zcorpan** [comment]
3 tests are failing locally for me in `openDialog3`, but they are failing in master as well. They seems to not be failing in Travis CI. Can anyone reproduce this? I couldn't figure out why this happens. Adding timeouts between steps or in `reload()` didn't help.

<details>
<summary>error log</summary>

```
3 tests failed

  dialog-modal_dialog › dialog-modal/dialog.html [data-test-id="key-tab"]: tab changes focus within dialog

  test/tests/dialog-modal_dialog.js:85

   84:   const dialog = (await t.context.queryElements(t, '#dialog3'))[0];
   85:   assert(                                                          
   86:     await dialog.isDisplayed(),                                    

  Rejected promise returned by test. Reason:

  AssertionError {
    actual: false,
    code: 'ERR_ASSERTION',
    expected: true,
    generatedMessage: false,
    operator: '==',
    message: 'dialog3 should have successfullly openned',
  }

  openDialog3 (test/tests/dialog-modal_dialog.js:85:3)
  test/tests/dialog-modal_dialog.js:249:3



  dialog-modal_dialog › dialog-modal/dialog.html [data-test-id="key-shift-tab"]: shift tab changes focus within dialog

  test/tests/dialog-modal_dialog.js:85

   84:   const dialog = (await t.context.queryElements(t, '#dialog3'))[0];
   85:   assert(                                                          
   86:     await dialog.isDisplayed(),                                    

  Rejected promise returned by test. Reason:

  AssertionError {
    actual: false,
    code: 'ERR_ASSERTION',
    expected: true,
    generatedMessage: false,
    operator: '==',
    message: 'dialog3 should have successfullly openned',
  }

  openDialog3 (test/tests/dialog-modal_dialog.js:85:3)
  test/tests/dialog-modal_dialog.js:358:3



  dialog-modal_dialog › dialog-modal/dialog.html [data-test-id="key-escape"]: escape closes dialog

  test/tests/dialog-modal_dialog.js:85

   84:   const dialog = (await t.context.queryElements(t, '#dialog3'))[0];
   85:   assert(                                                          
   86:     await dialog.isDisplayed(),                                    

  Rejected promise returned by test. Reason:

  AssertionError {
    actual: false,
    code: 'ERR_ASSERTION',
    expected: true,
    generatedMessage: false,
    operator: '==',
    message: 'dialog3 should have successfullly openned',
  }

  openDialog3 (test/tests/dialog-modal_dialog.js:85:3)
  test/tests/dialog-modal_dialog.js:439:5

```

</details>

**Comment by jongund** [comment]
@zcorpan 
Welcome to my world!

**Comment by jongund** [comment]
@zcorpan 

I ran the regression tests for "datepicker-dialog.html" on macOS and Window 10 with 34 passing tests and no failing tests.  So looking good for me.

**Comment by jongund** [comment]
There is a problem with this branch, it is modifying `combobox\combobox-datepicker.html` this needs to be removed before merging.  Combobox datepicker is modified in the [PR 1017](https://github.com/w3c/aria-practices/pull/1017) .

**Comment by zcorpan** [comment]
I've filed https://github.com/w3c/aria-practices/issues/1415 so we can track that test separately from this PR.

**Comment by jongund** [comment]
@mcking65 
The following changes have been completed based on your e-mail:
1. The aria-expanded has been removed form the Choose Date button.
2. Button by default has the label "Choose Date" and if there is a date in the texrbox the label is updated to “Change Date, value”. 
3. Merge conflicts have been removed.

**Comment by mcking65** [comment]
@jongund, I just wrapped up my editorial review and pushed some revisions. While reviewing, I found 2 more things we need to change before requesting final reviews.

1. Remove aria-autocomplete from the input. It is a normal textbox, not a combobox, so we shouldn't specify autocomplete none.
2. Remove support for down arrow opening the dialog when focus is on the button; modify script, test, and documentation accordingly. The button is a normal button, not a menu button, so it should not support activation with down arrow.

**Comment by jongund** [comment]
@mcking65 
Is the reason it is not to use the menu button pattern is because it opens a dialog and not a menu?

**Comment by jongund** [comment]
@mcking65 
The date picker dialog has been updated by removing the autocomplete attribute and the down arrow now does not open the dialog.

**Comment by mcking65** [comment]
@jongund wrote:

> Is the reason it is not to use the menu button pattern is because it opens a dialog and not a menu?

Yes, exactly.

**Comment by carmacleod** [comment]
The calendar button takes focus on page load. Do we want that? I found it a bit surprising.

**Comment by carmacleod** [comment]
> In the dialog, shortcut keys are assigned to the additional buttons for changing the month and year displayed in the calendar.

How can the user discover the shortcut keys? Should we have tooltips (@smhigley 😄 ?) on the change month/year buttons?

**Comment by carmacleod** [comment]
If the user only types a 2-digit year, what do we want to happen?
Currently, the calendar opens in the 1900's.
For example, 5/7/09 opens to May 7th, 1909.
I think we should at least change that to open May 7th, 2009, because now that we are in the 2000's, people think of '09 meaning 2009.

**Comment by smhigley** [comment]
@carmacleod could we open a separate issue for those bugs? I'm in favor of keeping this PR to the changes Jon already outlined & merging it (and those changes look good to me, fwiw). Then we can have additional smaller PRs for each other bug :).

**Comment by carmacleod** [comment]
@smhigley I'm happy to open new issues for the shortcuts and the 2-digit year. But the calendar button taking focus on page load is being introduced by this PR (here's the [released example](https://w3c.github.io/aria-practices/examples/dialog-modal/datepicker-dialog.html)), so I think we should fix that one before it goes in?
It seems to be because the DatePickerDialog is created on page load and then "this.close()" is called, but that puts focus on the button. So just needs a tiny refactor to call the "hide the dialog" code separately from the "focus the button" code.

**Comment by jongund** [comment]
@carmacleod 
@smhigley 
I fixed the focus on button when page loads and also added a feature for a 2 digit date to default to 20xx.
Please test when you have a chance.
Thank you for the feedback.

**Comment by jongund** [comment]
@carmacleod 
Thank you for all your testing and feedback.

**Pull request merged**